Use a strong type for weekday in GregorianDateTime

cdumez opened the pull request:
#### 61c45636fa179d656ac263df70a9ba9d1580d698
<pre>
Use a strong type for weekday in GregorianDateTime
<a href="https://bugs.webkit.org/show_bug.cgi?id=292682">https://bugs.webkit.org/show_bug.cgi?id=292682</a>

Reviewed by Darin Adler and Geoffrey Garen.

Use a strong type for weekday in GregorianDateTime instead of an int. It is
less error-prone and it removes ambiguity over whether 0 represents Monday
or Sunday. I wanted to use std::chrono::weekday but it is not yet available
in the playstation port&apos;s SDK. As a result, I am using an API-compatible
WTF::Weekday instead for now.

* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/runtime/DateConversion.cpp:
(JSC::formatDateTime):
* Source/JavaScriptCore/runtime/DatePrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/DateMath.cpp:
(WTF::weekdayName):
(WTF::makeRFC2822DateString):
* Source/WTF/wtf/DateMath.h:
(WTF::msToWeekDay):
(WTF::weekDay):
(WTF::equivalentYear):
* Source/WTF/wtf/GregorianDateTime.cpp:
(WTF::GregorianDateTime::setToCurrentLocalTime):
* Source/WTF/wtf/GregorianDateTime.h:

Canonical link: <a href="https://commits.webkit.org/294709@main">https://commits.webkit.org/294709@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/199e43a1a681bc9e28d2b550494827b6f0a48233

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102769 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22442 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12762 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107934 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53410 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22759 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30944 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78156 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35116 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105775 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92747 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58488 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10782 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52767 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/95444 "Built successfully and passed tests") | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10850 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110310 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101379 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29906 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22037 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30270 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88943 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86746 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22085 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31580 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9303 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24162 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29833 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35154 "Failed to build and analyze WebKit") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/125012 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29641 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34690 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32968 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31203 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/DateMath.cpp:
```diff
@@ -103,10 +103,16 @@ static Vector<UChar>& innerTimeZoneOverride() WTF_REQUIRES_LOCK(innerTimeZoneOve
+
+ASCIILiteral weekdayName(std::chrono::weekday weekday)
+{
+    static const std::array<ASCIILiteral, 7> weekdayName { "Sun"_s, "Mon"_s, "Tue"_s, "Wed"_s, "Thu"_s, "Fri"_s, "Sat"_s };
```
```suggestion
    static constexpr std::array<ASCIILiteral, 7> weekdayName { "Sun"_s, "Mon"_s, "Tue"_s, "Wed"_s, "Thu"_s, "Fri"_s, "Sat"_s };
```
Also not sure we need the `static` to generate efficient code if we have `constexpr`.

cdumez replied:
Based on sone quick online research, I think `static` is still useful for `constexpr` local variables.

darinadler commented on Source/WTF/wtf/DateMath.h:
```diff
@@ -373,18 +375,18 @@ inline int msToSeconds(double ms)
     if (wd < 0)
         wd += 7;
-    return wd;
+    return std::chrono::weekday { static_cast<unsigned>(wd) };
```
```suggestion
    return std::chrono::weekday { unsignedCast(wd) };
```

darinadler commented on Source/WTF/wtf/DateMath.h:
```diff
@@ -373,18 +375,18 @@ inline int msToSeconds(double ms)
 {
     int32_t result = (days + 4) % 7;
-    return result >= 0 ? result : result + 7;
+    return std::chrono::weekday { static_cast<unsigned>(result >= 0 ? result : result + 7) };
```
```suggestion
    return std::chrono::weekday { unsignedCast(result >= 0 ? result : result + 7) };
```

darinadler commented on Source/WTF/wtf/GregorianDateTime.cpp:
```diff
@@ -101,7 +101,7 @@ void GregorianDateTime::setToCurrentLocalTime()
     m_monthDay = localTM.tm_mday;
     m_yearDay = localTM.tm_yday;
-    m_weekDay = localTM.tm_wday;
+    m_weekDay = std::chrono::weekday { static_cast<unsigned>(localTM.tm_wday) };
```
```suggestion
    m_weekDay = std::chrono::weekday { unsignedCast(localTM.tm_wday) };
```

darinadler reviewed (approved):
Love this. Can consider using `unsignedCast` a bit too if you land your other pull request first.

cdumez committed:
Use a strong type for weekday in GregorianDateTime
https://bugs.webkit.org/show_bug.cgi?id=292682

Reviewed by Darin Adler and Geoffrey Garen.

Use a strong type for weekday in GregorianDateTime instead of an int. It is
less error-prone and it removes ambiguity over whether 0 represents Monday
or Sunday. I wanted to use std::chrono::weekday but it is not yet available
in the playstation port's SDK. As a result, I am using an API-compatible
WTF::Weekday instead for now.

* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/runtime/DateConversion.cpp:
(JSC::formatDateTime):
* Source/JavaScriptCore/runtime/DatePrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/DateMath.cpp:
(WTF::weekdayName):
(WTF::makeRFC2822DateString):
* Source/WTF/wtf/DateMath.h:
(WTF::msToWeekDay):
(WTF::weekDay):
(WTF::equivalentYear):
* Source/WTF/wtf/GregorianDateTime.cpp:
(WTF::GregorianDateTime::setToCurrentLocalTime):
* Source/WTF/wtf/GregorianDateTime.h:

Canonical link: https://commits.webkit.org/294709@main
